engine/engine.go

```diff
@@ -635,26 +664,158 @@ func (e *Engine) deleteKeyCidMap(ctx context.Context, provider peer.ID, contextI
 }
 
-type providerAndContext struct {
+func (e *Engine) removeProviderFromCidProviderAndKeyMap(ctx context.Context, c cid.Cid, providerIDBytes []byte) error {
+	pAndCBytes, err := e.ds.Get(ctx, e.cidToProviderAndKeyKey(c))
+
+	if err != nil && err != datastore.ErrNotFound {
+		return err
+	}
+	if err != datastore.ErrNotFound {
+		first := &ProviderAndContext{}
+		err = json.Unmarshal(pAndCBytes, first)
+		if err != nil {
+			return err
+		}
+
+		var prev *ProviderAndContext
+		curr := first
+
+		for {
+			if bytes.Equal(curr.Provider, providerIDBytes) {
+				if prev != nil {
+					prev.Next = curr.Next
+				}
+				if first == curr {
+					first = curr.Next
+				}
+			}
+			prev = curr
+			curr = curr.Next
+
+			if curr == nil {
+				break
+			}
+		}
+
+		if first == nil {
+			err = e.ds.Delete(ctx, e.cidToProviderAndKeyKey(c))
+			if err != nil {
+				return err
+			}
+		} else {
+			pAndCBytes, err = json.Marshal(first)
+			if err != nil {
+				return err
+			}
+			err = e.ds.Put(ctx, e.cidToProviderAndKeyKey(c), pAndCBytes)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (e *Engine) deleteAllMapsForProvider(ctx context.Context, provider peer.ID) error {
+	// cleaning up index for provider works as follows
+	// - scanning the datastore for all keys starting with either map/keyCid/ or map/keyCid/providerId for non default providers
+	// - for each key removing cidKey, cidProviderAndKey, keyCid records
+	// - performing another index scan on keyMetadataIndex and cleaning up related records form there
+
+	keyToCidPrefix := keyToCidMapPrefix
+	metadataPrefix := keyToMetadataMapPrefix
+	if provider != e.provider.ID {
+		keyToCidPrefix += provider.String()
+		metadataPrefix += provider.String()
+	}
+	pB, err := provider.Marshal()
+	if err != nil {
+		return err
+	}
+	err = e.deleteAllKeysForQuery(ctx, &query.Query{
+		Prefix: keyToCidPrefix,
+	}, func(r query.Result) error {
+		_, c, err := cid.CidFromBytes(r.Value)
+		if err != nil {
+			return err
+		}
+		// cleaning up the legacy index of cid -> contextID
+		err = e.ds.Delete(ctx, e.cidToKeyKey(c))
+		if err != nil {
+			return err
+		}
+
+		// removing all provider related records from the indexed linked lists
+		err = e.removeProviderFromCidProviderAndKeyMap(ctx, c, pB)
+		if err != nil {
+			return err
+		}
+
+		return e.ds.Delete(ctx, datastore.NewKey(r.Key))
+	})
+
+	if err != nil {
+		return err
+	}
+
+	// deleting key to metadata mappings
+	return e.deleteAllKeysForQuery(ctx, &query.Query{
+		Prefix:   metadataPrefix,
+		KeysOnly: true,
+	}, func(r query.Result) error {
+		return e.ds.Delete(ctx, datastore.NewKey(string(r.Key)))
+	})
+}
+
+func (e *Engine) deleteAllKeysForQuery(ctx context.Context, q *query.Query, f func(query.Result) error) error {
+	results, err := e.ds.Query(ctx, *q)
+	if err != nil {
+		return err
+	}
+	defer results.Close()
+
+	for r := range results.Next() {
+		if r.Error != nil {
+			return r.Error
+		}
+
+		err = f(r)
```

Comment:
I don't understand why we need to pass a function `f` here. Isn't deletion always the same?